lib/bolt/cli.rb

```diff
@@ -379,725 +337,570 @@ def validate(options)
+    # @param options [Hash] The CLI options.
+    #
+    private def process_command(app, command, action, options)
+      case command
+      when 'apply'
+        results = outputter.spin do
+          app.apply(options[:object], options[:targets], **options.slice(:code, :noop))
+        end
+        rerun.update(results)
+        app.shutdown
+        outputter.print_apply_result(results)
+        results.ok? ? SUCCESS : FAILURE
 
-      if %w[task plan].include?(options[:subcommand]) && options[:task_options] && !options[:params_parsed] && pal
-        options[:task_options] = pal.parse_params(options[:subcommand], options[:object], options[:task_options])
-      end
+      when 'command'
+        outputter.print_head
+        results = app.run_command(options[:object], options[:targets], **options.slice(:env_vars))
+        rerun.update(results)
+        app.shutdown
+        outputter.print_summary(results, results.elapsed_time)
+        results.ok? ? SUCCESS : FAILURE
+
+      when 'file'
+        case action
+        when 'download'
+          outputter.print_head
+          results = app.download_file(options[:object], options[:leftovers].first, options[:targets])
+          rerun.update(results)
+          app.shutdown
+          outputter.print_summary(results, results.elapsed_time)
+          results.ok? ? SUCCESS : FAILURE
+        when 'upload'
+          outputter.print_head
+          results = app.upload_file(options[:object], options[:leftovers].first, options[:targets])
+          rerun.update(results)
+          app.shutdown
+          outputter.print_summary(results, results.elapsed_time)
+          results.ok? ? SUCCESS : FAILURE
+        end
+
+      when 'group'
+        outputter.print_groups(**app.list_groups)
+        SUCCESS
 
-      case options[:subcommand]
       when 'guide'
-        code = if options[:object]
-                 show_guide(options[:object])
-               else
-                 list_topics
-               end
-      when 'project'
-        case options[:action]
-        when 'init'
-          code = Bolt::ProjectManager.new(config, outputter, pal)
-                                     .create(Dir.pwd, options[:object], options[:modules])
-        when 'migrate'
-          code = Bolt::ProjectManager.new(config, outputter, pal).migrate
+        if options[:object]
+          outputter.print_guide(**app.show_guide(options[:object]))
+        else
+          outputter.print_topics(**app.list_guides)
         end
-      when 'lookup'
-        plan_vars = Hash[options[:leftovers].map { |a| a.split('=', 2) }]
-        # Validate functions verifies one of these was passed
-        if options[:targets]
-          code = lookup(options[:object], options[:targets], plan_vars: plan_vars)
-        elsif options[:plan_hierarchy]
-          code = plan_lookup(options[:object], plan_vars: plan_vars)
+        SUCCESS
+
+      when 'inventory'
+        targets = app.show_inventory(options[:targets])
```

Comment:
`options[:targets]` will actually be `[]` here if no targetting options are provided, not `nil`, which means `bolt inventory show` is actually always empty...I think checking for `empty?` instead below makes sense, but for this maybe something like:
`options[:targets].empty? ? nil : options[:targets]`

Also vaguely concerning we don't have a test that caught this....

Reply:
Also verified this is the only place we assume that `options[:targets]` can be nil.

Reply:
Fixed and added an integration test for this. :)